types/handler.go

```diff
@@ -31,13 +31,25 @@ func ChainAnteDecorators(chain ...AnteDecorator) AnteHandler {
 		return nil
 	}
 
-	// handle non-terminated decorators chain
-	if (chain[len(chain)-1] != Terminator{}) {
-		chain = append(chain, Terminator{})
-	}
+	noOpDecorator := func(ctx Context, _ Tx, _ bool) (Context, error) { return ctx, nil }
 
 	return func(ctx Context, tx Tx, simulate bool) (Context, error) {
-		return chain[0].AnteHandle(ctx, tx, simulate, ChainAnteDecorators(chain[1:]...))
+		var err error
+
+		// We call each AnteHandler in the chain in sequential iterative order,
+		// where the next AnteHandler call is a no-op decorator. We pass the resulting
+		// Context of each call to the next AnteHandler.
+		//
+		// If a recursive approach is used instead, then the next AnteHandler must
+		// be a valid decorator except for the last which should be a Terminator.
+		for _, d := range chain {
+			ctx, err = d.AnteHandle(ctx, tx, simulate, noOpDecorator)
+			if err != nil {
+				return ctx, err
+			}
+		}
```

Comment:
Ahhh, right. Since the recursive approach executes the chain in single call stack, any panic in a decorator will be caught be `Setup`.

In the iterative approach, a triggered panic will bubble up to the function closure but will not be recovered as it is in the recursive approach.

I think what we can do is:

- Remove `SetUpContextDecorator`
- Move the existing `SetUpContextDecorator` logic (you've highlighted) directly to the function closure in `ChainDecorators`. WDYT?

Reply:
That SGTM!

Reply:
Thanks for eliminating one more "crazy design edge case" from our brain caches <3